### PR TITLE
De-register forward hooks upon exiting flop counter context

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103744

This PR fixes https://github.com/pytorch/pytorch/issues/103684.
- Instead of registering forward hooks in `__init__()`, do it upon `__enter__()`.
- De-register those forward hooks upon `__exit__()`.
- Achieve this by saving an additional mapping `_module_to_forward_hook_handles: Dict[nn.Module, _ForwardHookHandles]`. Only the values in the mapping (i.e. not the keys) are useful for this change. (A `List[_ForwardHookHandles]` would suffice.)
- The unit test accesses private attributes `_forward_hooks` and `_forward_pre_hooks` :/

Note that this PR is technically not backward compatible since it does not register the hooks upon `__init__()`, which means that you will not get the flops counting without the context manager.

### CODE REVIEW

File: torch/utils/flop_counter.py [hunk 1]
```diff
@@ -244,18 +246,32 @@ def __init__(
...
+        # Keys will include the modules in `mods` and their submodules
+        self._module_to_forward_hook_handles: Dict[nn.Module, _ForwardHookHandles] = {}
         self.flop_mapping = {**flop_mapping, **custom_mapping}
 
+    def _register_forward_hooks(self):
+        if self.mods is None:
+            return
+        for mod in self.mods:
+            prefix = type(mod).__name__
+            for name, module in dict(mod.named_modules()).items():
```
Comment (awgu) [inline torch/utils/flop_counter.py:258]: minor: I copied this over, but I was wondering why not just `for name, module in mod.named_modules():`?

Comment (Chillee) [reply torch/utils/flop_counter.py:258]: Uhh... no particular reason I think.

**Review by Chillee — APPROVED** [review]
LGTM! Thanks

**Pull request closed without merging**